src/adaptor/proxy/command.ts

```diff
@@ -0,0 +1,181 @@
+import {
+  type APIActionRowComponent,
+  type APIMessageActionRowComponent,
+  ActionRowBuilder,
+  ButtonBuilder,
+  ButtonStyle,
+  Client,
+  Message,
+  type MessageActionRowComponentBuilder
+} from 'discord.js';
+import type {
+  CommandProxy,
+  MessageCreateListener
+} from '../../runner/command.js';
+import { Schema, makeError } from '../../model/command-schema.js';
+import type { EmbedPage } from '../../model/embed-message.js';
+import type { RawMessage } from './middleware.js';
+import type { Snowflake } from '../../model/id.js';
+import { convertEmbed } from '../embed-convert.js';
+import { parseStrings } from './command/schema.js';
+
+const SPACES = /\s+/;
+
+export class DiscordCommandProxy implements CommandProxy {
+  constructor(client: Client, private readonly prefix: string) {
+    client.on('messageCreate', (message) => this.onMessageCreate(message));
+  }
+
+  private readonly listenerMap = new Map<
+    string,
+    [Schema, MessageCreateListener]
+  >();
+
+  addMessageCreateListener(
+    schema: Schema,
+    listener: MessageCreateListener
+  ): void {
+    for (const name of schema.names) {
+      if (this.listenerMap.has(name)) {
+        throw new Error(`command name conflicted: ${name}`);
+      }
+      this.listenerMap.set(name, [schema, listener]);
+    }
+  }
+
+  private async onMessageCreate(message: Message): Promise<void> {
+    if (message.author.bot || message.author.system) {
+      return;
+    }
+    await message.fetch();
+
+    if (!message.content?.trimStart().startsWith(this.prefix)) {
+      return;
+    }
+    const args = message.content
+      ?.trim()
+      .slice(this.prefix.length)
+      .split(SPACES);
+
+    const entry = this.listenerMap.get(args[0]);
+    if (!entry) {
+      return;
+    }
+    const [schema, listener] = entry;
+    const [tag, parsedArgs] = parseStrings(args, schema);
+    if (tag === 'Err') {
+      const error = makeError(parsedArgs);
+      await message.reply(error.message);
+      return;
+    }
+
+    await listener({
+      senderId: message.author.id as Snowflake,
+      senderGuildId: message.guildId as Snowflake,
+      senderChannelId: message.channelId as Snowflake,
+      get senderVoiceChannelId(): Snowflake | null {
+        const id = message.member?.voice.channelId ?? null;
+        return id ? (id as Snowflake) : null;
+      },
+      senderName: message.author?.username ?? '名無し',
```

Comment:
```suggestion
      senderName: message.author?.username ?? '取得エラー',
```

これ名無しよりも取得エラーのほうが望ましいと思うんですけど、どうですかね

Reply:
これは元の仕様をそのまま引き継いでいます. これを変えると小さい破壊的変更になると思われます. また, メッセージの作成者が存在しないことは *取得エラー* とは限らないでしょう.